core/src/sumeragi/fault.rs

```diff
@@ -38,866 +37,1337 @@ impl FaultInjection for NoFault {
-    /// Initializes sumeragi with the `latest_block_hash` and `block_height` after Kura loads the blocks.
-    pub fn init(&mut self, latest_block: HashOf<VersionedCommittedBlock>, block_height: u64) {
-        self.block_height = block_height;
-        self.topology.apply_block(latest_block);
-    }
-
-    /// Updates network topology by taking the actual list of peers from `WorldStateView`.
-    /// Updates it only if there is a change in WSV peers, otherwise leaves the order unchanged.
-    #[allow(clippy::expect_used, clippy::unused_async)]
-    pub async fn update_network_topology(&mut self) {
-        let wsv_peers: HashSet<_> = self.wsv.trusted_peers_ids().clone().into_iter().collect();
-        let topology_peers: HashSet<_> = self.topology.sorted_peers().iter().cloned().collect();
-        if topology_peers != wsv_peers {
-            self.topology = self.topology
-                .clone()
-                .into_builder()
-                .with_peers(wsv_peers)
-                .build()
-                // TODO: Check it during instruction execution.
-                .expect("The safety of changing the number of peers should have been checked at Instruction execution stage.");
-        }
-    }
-
-    /// Returns `true` if some block is in discussion, `false` otherwise.
-    #[allow(clippy::unused_async)]
-    pub async fn voting_in_progress(&self) -> bool {
-        self.voting_block.is_some()
-    }
-
-    /// Latest block hash as seen by sumeragi.
-    pub fn latest_block_hash(&self) -> &HashOf<VersionedCommittedBlock> {
-        self.topology.at_block()
-    }
-
-    /// Number of view changes.
-    /// Where a view change is a change in topology made if there was some consensus misfunction during round (faulty peers).
-    pub fn number_of_view_changes(&self) -> u64 {
-        self.topology.view_change_proofs().len() as u64
-    }
-
-    /// The proofs of view changes that happened after the last block was committed.
-    pub fn view_change_proofs(&self) -> &ViewChangeProofs {
-        self.topology.view_change_proofs()
-    }
-
-    /// The hash of the latest view change.
-    pub fn latest_view_change_hash(&self) -> HashOf<Proof> {
-        self.view_change_proofs().latest_hash()
-    }
+#[allow(clippy::expect_used)]
+fn request_view_change<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    state_machine: &mut SumeragiStateMachineData,
+    view_change_proof_chain: &mut Vec<Proof>,
+    current_view_change_index: u64,
+) where
+    F: FaultInjection,
+{
+    let suspect_proof = {
+        let mut proof = Proof {
+            latest_block_hash: state_machine.latest_block_hash,
+            view_change_index: current_view_change_index,
+            signatures: Vec::new(),
+        };
+        proof
+            .sign(sumeragi.key_pair.clone())
+            .expect("Suspect proof must be able to perform signing");
+        proof
+    };
+
+    let _ = view_change_proof_chain.insert_proof(
```

Comment:
We are ignoring error here and then broadcast probably wrong `Vec<Proof>` to other peers, is it ok?

Reply:
Originally I didn't have a return type because the code can't 'fail'. Aleksandr wanted to have a return type to know what was going on inside. `insert_proof` is written such that you can insert any/all possible proofs and only the correct ones will stick. It internally combines the proof with the existing proofs where possible and removes redundant signatures et cetera.

Reply:
I.e. we need to check for the error message, and hope that the designers of the language were not idiots and marked the error path as the cold path.